Add polling to power functions

britthouser opened the pull request:
Update the power up/down functions to poll the server to ensure the desired power state is reached.  Also misc flake8 compliance fixes.

britthouser committed:
Fix flake8 errors in serveractions.py

Three lines in serveractions.py which were longer than 79
chars and caused flake8 to fail.  Split those lines so
flake8 passes now.

britthouser committed:
Add polling to power up/down functions

This patch adds logic to the power up/down functions in the apis
folder to poll the server and return only after the desired power
state has happened been reached.

Signed-off-by: Britt Houser <bhouser@cisco.com>

vvb commented on imcsdk/apis/server/serveractions.py:
```diff
@@ -46,21 +49,42 @@ def power_up_server(handle):
+        handle.set_mo(rack_mo)
+
+    # Poll until the server is powered up
+    wait_time = 0
```
@britthouser This entire wait logic is sort of similar for power_up and power_down except for the state we have to check on.. can we abstract this in a internal method in the same file, `_wait_for_power_state(handle, "on"/"off")` and move all this logic there. some De-dup.

britthouser replied:
Done!

vvb commented on imcsdk/apis/server/serveractions.py:
```diff
@@ -46,21 +49,42 @@ def power_up_server(handle):
+    while get_server_power_state(handle) != "on":
+        # Raise error if we've reached timeout
+        if wait_time > timeout:
+            raise RuntimeError(
```
@britthouser we should use ImcOperationError here..

```
raise ImcOperationError("String describing Operation", "String describing Error")
```

Look in imcsdk/apis/server/adaptor.py for examples.

britthouser replied:
Done!

vvb commented on imcsdk/apis/server/serveractions.py:
```diff
@@ -69,13 +93,30 @@ def power_down_server(handle):
-    handle.set_mo(rack_mo)
-    return rack_mo
+    # Turn power off only if not already powered down
+    if get_server_power_state(handle) != "off":
```
same as above.. De-dup

britthouser replied:
Done!

vvb commented on imcsdk/apis/server/serveractions.py:
```diff
@@ -69,13 +93,30 @@ def power_down_server(handle):
+    while get_server_power_state(handle) != "off":
+        # Raise error if we've reached timeout
+        if wait_time > timeout:
+            raise RuntimeError(
```
same as above.. move to `ImcOperationError`

britthouser replied:
Done!

vvb commented on imcsdk/apis/server/serveractions.py:
```diff
@@ -46,22 +49,42 @@ def power_up_server(handle):
+    wait_time = 0
+    while get_server_power_state(handle) != "on":
+        # Raise error if we've reached timeout
+        if wait_time > timeout:
```
@britthouser For accuracy this should be `wait_time >= timeout`, otherwise we would end up waiting for another $interval duration because of the sleep.

britthouser replied:
I used > so that if the timeout is set to zero (e.g. in unit testing) then we go through the loop one time.

vvb commented on imcsdk/apis/server/serveractions.py:
```diff
@@ -70,13 +93,30 @@ def power_down_server(handle):
+    wait_time = 0
+    while get_server_power_state(handle) != "off":
+        # Raise error if we've reached timeout
+        if wait_time > timeout:
```
`>=`

britthouser replied:
Same as above.

vvb reviewed (changes requested):
@britthouser thanks for the PR. The changes look great, except for some minor nits. 

Regarding the tests, I see that we are missing some testing for existing code as well. So, I will take care of adding the tests in a separate PR.

vvb commented:
also pls squash the commits.

britthouser committed:
Refactor server polling

The server polling loop is identical between functions except
for the state its watching for.  So refactored the loop into
its own function that takes in a state.

Signed-off-by: Britt Houser <bhouser@cisco.com>

britthouser committed:
Add polling to graceful power down and power cycle

Added the polling loop to both the graceful power
down and the power cycle functions.

Signed-off-by: Britt Houser <bhouser@cisco.com>

britthouser committed:
Address first round of commends in PR

Changed the error raised to be a ImcOperationError.

Signed-off-by: Britt Houser <bhouser@cisco.com>

britthouser commented:
Sorry for my github ignorance (I've only ever used gerrit), but I'm not sure how to squash the commits?
